rest/src/main/scala/io/udash/rest/raw/RestRequest.scala

```diff
@@ -0,0 +1,44 @@
+package io.udash
+package rest
+package raw
+
+import com.avsystem.commons._
+import com.avsystem.commons.meta._
+import com.avsystem.commons.misc.{AbstractValueEnum, AbstractValueEnumCompanion, EnumCtx}
+import com.avsystem.commons.rpc._
+
+import scala.util.control.NoStackTrace
+
+/**
+  * Enum representing HTTP methods.
+  */
+final class HttpMethod(implicit enumCtx: EnumCtx) extends AbstractValueEnum
+object HttpMethod extends AbstractValueEnumCompanion[HttpMethod] {
+  final val GET, HEAD, POST, PUT, DELETE, CONNECT, OPTIONS, TRACE, PATCH: Value = new HttpMethod
+}
+
+/**
+  * Aggregates serialized path, header and query parameters of an HTTP request.
+  */
+case class RestParameters(
+  @multi @tagged[Path] path: List[PlainValue] = Nil,
+  @multi @tagged[Header] headers: IMapping[PlainValue] = IMapping.empty,
+  @multi @tagged[Query] query: Mapping[PlainValue] = Mapping.empty
+) {
+  def append(method: RestMethodMetadata[_], otherParameters: RestParameters): RestParameters =
+    RestParameters(
+      path ::: method.applyPathParams(otherParameters.path),
+      headers ++ otherParameters.headers,
+      query ++ otherParameters.query
+    )
+}
+object RestParameters {
+  final val Empty = RestParameters()
+}
+
+case class HttpErrorException(code: Int, payload: OptArg[String] = OptArg.Empty, cause: Throwable = null)
```

Comment:
final

Reply:
I could actually see some benefit in this being non-final.